Saving Princess: implement new game

LeonarthCG opened the pull request:
## What is this fixing or adding?

Adds support for Saving Princess to Archipelago.

## How was this tested?

The world was tested with an async that had 3 players run the game, together with another 6 games. Only two major issues were found during this, which have since been fixed.

Over the past month and a half I have also played seeds by myself, testing every option by running the game on its own, with other slots of the game and also with other supported and unsupported games.

## If this makes graphical changes, please attach screenshots.
![image](https://github.com/ArchipelagoMW/Archipelago/assets/33758848/c99260bf-06af-4165-a1ef-4ddb957ebadc)

ScipioWright commented on worlds/saving_princess/__init__.py:
```diff
@@ -0,0 +1,123 @@
+    def get_filler_item_name(self) -> str:
+        filler_list = list(Items.item_dict_filler.keys())
+        # check if this is going to be a trap
+        if self.multiworld.random.randint(0, 99) < self.options.trap_chance:
+            filler_list = list(Items.item_dict_traps.keys())
+        # and return one of the names at random
+        return self.multiworld.random.choice(filler_list)
```
```suggestion
        if self.random.randint(0, 99) < self.options.trap_chance:
            filler_list = list(Items.item_dict_traps.keys())
        # and return one of the names at random
        return self.random.choice(filler_list)
```

Would recommend using the world's random object instead of the multiworld random object. Makes it more deterministic, basically.

ScipioWright commented on worlds/saving_princess/docs/en_Saving Princess.md:
```diff
@@ -0,0 +1,74 @@
+# Saving Princess
+
+## Where is the settings page?
```
```suggestion
## Where is the options page?
```

ScipioWright commented on worlds/saving_princess/docs/setup_en.md:
```diff
@@ -0,0 +1,134 @@
+
+### Where do I get a YAML file?
+
+You can customize your settings by visiting the [Saving Princess Player Settings Page](/games/Saving Princess/player-settings)
```
```suggestion
You can customize your options by visiting the [Saving Princess Player Options Page](/games/Saving Princess/player-options)
```

Settings was changed to Options a while back

ScipioWright commented on worlds/saving_princess/Regions.py:
```diff
@@ -0,0 +1,126 @@
+from .Constants import *
+
+
+class RegionData:
+    location_names: List[str]
+
+    def __init__(self, location_names: List[str]):
+        self.location_names = location_names
+
+
+region_dict: Dict[str, RegionData] = {
+    REGION_MENU: RegionData([]),
+    REGION_CAVE: RegionData(
+        [
+            LOCATION_CAVE_AMMO,
```
Is there a specific reason this is a `Dict[str, RegionData]` rather than a `Dict[str, List[str]]`? The surrounding class feels unnecessary, but maybe I'm just missing something here.

LeonarthCG replied:
When I first made the class there was more to it, but you are right that as it currently stands there's no reason for it to exist. I'll do away with it.

ScipioWright commented on worlds/saving_princess/Regions.py:
```diff
@@ -0,0 +1,126 @@
+    ]
+
+
+def create_regions(world: MultiWorld, player: int, is_pool_expanded: bool):
+    for region_name, region_data in region_dict.items():
+        region = Region(region_name, player, world)
+        set_region_locations(region, region_data.location_names, is_pool_expanded)
+        world.regions.append(region)
+    connect_regions(world, player)
```
```suggestion
def create_regions(multiworld: MultiWorld, player: int, is_pool_expanded: bool):
    for region_name, region_data in region_dict.items():
        region = Region(region_name, player, multiworld)
        set_region_locations(region, region_data.location_names, is_pool_expanded)
        multiworld.regions.append(region)
    connect_regions(multiworld, player)
```

Please do not use `world: MultiWorld`. It can get confusing, both for people reviewing the code, and people using it as a reference while developing their own apworlds.

ScipioWright reviewed (commented):
Just a quick skim, looks pretty great so far. Will review it in more depth later.

ScipioWright commented on worlds/saving_princess/Rules.py:
```diff
@@ -0,0 +1,110 @@
+        return (
+                state.has(ITEM_RELOAD_SPEED, world.player, 4)
+                and (
+                        state.has(ITEM_WEAPON_FIRE, world.player)
+                        or state.has(ITEM_WEAPON_ICE, world.player)
+                        or state.has(ITEM_WEAPON_VOLT, world.player)
```
```suggestion
                        state.has_any({ITEM_WEAPON_FIRE, ITEM_WEAPON_ICE, ITEM_WEAPON_VOLT}, world.player)
```
You can use state.has_any to do basically what you're doing here

ScipioWright commented on worlds/saving_princess/Rules.py:
```diff
@@ -0,0 +1,110 @@
+        if world.is_pool_expanded:
+            # in expanded, the final area requires all the boss keys
+            return (
+                    state.has(EP_ITEM_GUARD_GONE, world.player)
+                    and state.has(EP_ITEM_CLIFF_GONE, world.player)
+                    and state.has(EP_ITEM_ACE_GONE, world.player)
+                    and state.has(EP_ITEM_SNAKE_GONE, world.player)
```
```suggestion
                    state.has_all({EP_ITEM_GUARD_GONE, EP_ITEM_CLIFF_GONE, EP_ITEM_ACE_GONE, EP_ITEM_SNAKE_GONE}, world.player)
```

You can use state.has_all to do the same thing you had here

ScipioWright commented on worlds/saving_princess/Options.py:
```diff
@@ -0,0 +1,119 @@
+    display_name = "Music Shuffle"
+
+
+@dataclass
+class SavingPrincessOptions(PerGameCommonOptions):
+    death_link: DeathLink
+    start_inventory_from_pool: StartInventoryPool
+    expanded_pool: ExpandedPool
+    instant_saving: InstantSaving
+    sprint_availability: SprintAvailability
+    cliff_weapon_upgrade: CliffWeaponUpgrade
+    ace_weapon_upgrade: AceWeaponUpgrade
+    shake_intensity: ScreenShakeIntensity
+    iframes_duration: IFramesDuration
+    trap_chance: TrapChance
+    music_shuffle: MusicShuffle
```
```suggestion
@dataclass
class SavingPrincessOptions(PerGameCommonOptions):
    start_inventory_from_pool: StartInventoryPool
    expanded_pool: ExpandedPool
    trap_chance: TrapChance
    death_link: DeathLink
    instant_saving: InstantSaving
    sprint_availability: SprintAvailability
    cliff_weapon_upgrade: CliffWeaponUpgrade
    ace_weapon_upgrade: AceWeaponUpgrade
    shake_intensity: ScreenShakeIntensity
    iframes_duration: IFramesDuration
    music_shuffle: MusicShuffle
```

The order of the options in your dataclass determines what order the options show up in the player yamls and what order they show up in on the website. I recommend grouping the "generation" options together towards the top and the "client" options towards the bottom. The suggestion is my recommendation here.

ScipioWright reviewed (approved):
Small game, the apworld is written very cleanly. Unit tests passed fine, no errors when genning lots of random seeds. I did not test the client at all.

nicholassaylor reviewed (changes requested):
# Nature of Review
* Only `Options.py` and the documentation was reviewed
* Documentation was checked through an editor and verified through WebHost

### Options
Looking over the options, I did not see anything of concern, the tooltips were self-explanatory and the types made sense.

### Documentation
Documentation, particularly in the game page is fairly good. I left some comments that I think will make the pages a bit more readable and condense information.

### Client
Looking over the set-up guide, I see that a `.bs4diff` file is used to patch the `data.win` file of a Gamemaker game. I see that there are concerns about assets and code being included, but you may want to look at the Undertale implementation for patching or AP Procedure Patch (APPP) to see if you can integrate the patching process into either a launcher for patching or a standalone program. I don't consider this a requirement for merging, but it may make the game more approachable for randomizing.

### Additional Housekeeping
The following things should also get done before merging
* Add game to `README.md`
* Add your implementation to `CODEOWNERS`

ScipioWright commented on worlds/saving_princess/__init__.py:
```diff
@@ -0,0 +1,123 @@
+
+from BaseClasses import Tutorial, ItemClassification as ItemClass
+from Options import PerGameCommonOptions
+from ..AutoWorld import World, WebWorld
```
```suggestion
from worlds.AutoWorld import World, WebWorld
```

Imports from outside your world need to be absolute per the specification here: https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/apworld%20specification.md#caveats

Exempt-Medic commented on worlds/saving_princess/Items.py:
```diff
@@ -0,0 +1,92 @@
+        if code is not None:
+            self.code += BASE_ID
+
+        if self.item_class == ItemClass.filler or code is None:
```
I don't believe this `if` block ever does anything because all your filler and victory items explicitly set their `count` to 0 and the default for `count_extra` is already 0

LeonarthCG replied:
I agree, it is redundant and I should decide on one or the other.

I personally think that keeping the `if` block (and adding the check for `ItemClass.trap` that it is missing) while removing the redundant arguments in the item data dicts would look cleaner.

LeonarthCG commented on worlds/saving_princess/docs/en_Saving Princess.md:
```diff
@@ -0,0 +1,74 @@
+
+In Expanded Pool, the way to open the final area of the game changes, going from having to defeat bosses to having to locate some items, which  can also alter the order the game is tackled in.
+
+Finally, there are also new items to work as filler and traps, ranging from a full health and ammo restore to spawning a Ninja on top of you.
```
I have read the different comments you gave me and tried to rewrite the info page so that it is more condensed and information is not duplicated.

nicholassaylor commented on README.md:
This has a merge conflict and needs to be resolved

nicholassaylor commented on SavingPrincessClient.py:
I believe that in general, core is trying to shift away from clients in the root AP directory and instead having them open through the launcher.

nicholassaylor commented on worlds/saving_princess/docs/setup_en.md:
```diff
@@ -0,0 +1,120 @@
+
+1. Purchase and download [Saving Princess](https://brainos.itch.io/savingprincess)
+2. Download and install the latest [Archipelago release](https://github.com/ArchipelagoMW/Archipelago/releases/latest)
+3. Launch `ArchipelagoLauncher` or `ArchipelagoSavingPrincessClient` from your Archipelago installation folder
+4. If you launched `ArchipelagoLauncher`, click on "Saving Princess Client"
+   * You will probably need to scroll down on the Clients column to see it
```
If you end up moving the client to launcher only, this would need to be changed

nicholassaylor commented on worlds/saving_princess/docs/setup_en.md:
```diff
@@ -0,0 +1,120 @@
+
+### Where do I get a YAML file?
+
+You can customize your options by visiting the [Saving Princess Player Options Page](/games/Saving Princess/player-options).
```
```suggestion
You can customize your options by visiting the [Saving Princess Player Options Page](/games/Saving%20Princess/player-options).
```
Fix broken link

nicholassaylor commented on worlds/saving_princess/Client.py:
```diff
@@ -0,0 +1,215 @@
+    os.chdir(SavingPrincessWorld.settings.install_folder)
+
+    # check that the mod installation is valid
+    valid_install = is_install_valid()
+    if not valid_install:
```
```suggestion
    if not is_install_valid():
```
No need to hold onto this value since it's not used elsewhere

nicholassaylor reviewed (approved):
# Nature of Review
* All aspects of previous review were re-reviewed
* Client was briefly looked over

### Documentation
Changes look good, very minor changes suggested, otherwise very good

### Client
Client now has an automated patch feature while still retaining the option to patch manually. I only took a cursory look over the patching process and didn't see any glaring issues. I was not able to test patching as I do not own the game.

### Options
No changes since last review, however with #2614 being merged after my last review, it may be a good idea to take a look at Option Groups and Presets before merging. Not a requirement, but looking at the dataclass for your options, it seems like Option Groups would serve you well.

### Housekeeping
Game was added to `README.md` and `CODEOWNERS` but a merge conflict on `README.md` needs to be resolved. Minor inconvenience at most.

Overall, the concerns that I had previously were addressed and any remaining reservations are extremely minor and/or optional.

Exempt-Medic commented on worlds/saving_princess/__init__.py:
```diff
@@ -0,0 +1,156 @@
+        # and return one of the names at random
+        return self.random.choice(filler_list)
+
+    def fill_slot_data(self) -> Dict[str, Any]:
```
You could use `self.options.as_dict("death_link", "expanded_pool" "instant_saving", "sprint_availability", "cliff_weapon_upgrade", "ace_weapon_upgrade", "shake_intensity", "iframes_duration")` instead here.

Exempt-Medic commented on worlds/saving_princess/__init__.py:
```diff
@@ -0,0 +1,156 @@
+            "music_table": self.music_table,
+        }
+
+    def set_rules(self):
+        from .Rules import set_rules
+        set_rules(self)
```
This should probably be moved to line up with when it occurs in the call order, namely after `create_items` / before `fill_slot_data`

Exempt-Medic commented on worlds/saving_princess/Rules.py:
```diff
@@ -0,0 +1,120 @@
+from .Constants import *
+
+
+def set_rules(world: "SavingPrincessWorld"):
```
Do you have to import `SavingPrincessWorld` then?

LeonarthCG replied:
Not sure if you mean I'm missing the import (I think this is what other worlds do/what I'm missing):
```py
from typing import TYPE_CHECKING
if TYPE_CHECKING:
  from . import SavingPrincessWorld
```
Or if you are asking why I'm using it in the first place?

Exempt-Medic replied:
Yeah, I'm saying you'd want to include that exact code here

Exempt-Medic reviewed (changes requested):
A lot of minor things and mostly docs. Overall this all looks very good, though some of the helper functions, like `get_location` confuse me. Only major thing was that you're putting events in the `name_to_id` lookups. I looked at everything except `Client.py` and have not done any generations yet

Exempt-Medic commented:
I've now done one thousand generations (wow this game gens _fast_) and no errors occurred across 250 solo gens and 750 gens of increasing player size. Went over the WebHost display and everything looks fine there as well, you could potentially consider adding option groups or option presets, but it's small enough that it seems alright without.

Exempt-Medic commented:
Having now played the game, it connects and seems to run just fine, however you're connecting with the `AP` tag which you likely shouldn't be doing. You also may want to add code in `generate_early` to check that the player name can actually be input into your game. Could do something like:
```py
from Options import OptionError, PerGameCommonOptions

    def generate_early(self) -> None:
        if not self.player_name.isascii():
            raise OptionError(f"{self.player_name}'s name must be only ASCII.")
```

Exempt-Medic commented on worlds/saving_princess/Items.py:
```diff
@@ -0,0 +1,98 @@
+            self.count = count
+            self.count_extra = count_extra
+
+    def create_item(self, player: int):
```
Instead of creating these every single time you create an item, why not just make lookup tables directly beforehand that are shared and then use those?

LeonarthCG replied:
Sorry, I'm not sure I follow.
I think you mean `count` and `count_extra`? But I'm pretty sure those are only touched once per entry in my item dictionary, which is the only place `ItemData` instances are made (unless I made a mistake somewhere).

Or maybe it's my class naming that is confusing? `ItemData` holds data used to create `SavingPrincessItem` instances, but the `SavingPrincessItem` class doesn't have `ItemData` in it.

Exempt-Medic replied:
I'm saying that every time you call `create_item` the `index` and `name` variables get totally rebuilt. But since the item dict is static, you could just create these once, right after making the item dict itself instead of creating them many times

LeonarthCG replied:
I understand now. Maybe something like this would make more sense:
```py
    def create_item(self, player: int):
        return SavingPrincessItem(item_data_names[self], self.item_class, self.code, player)

[... all of the ItemData dict stuff ...]

item_data_names: Dict[ItemData, str] = {value: key for key, value in item_dict.items()}
````

Exempt-Medic replied:
Looks good to me

LeonarthCG commented on worlds/saving_princess/docs/en_Saving Princess.md:
```diff
@@ -0,0 +1,56 @@
+This textbox shows both which item you got and which player sent it to you.
+
+If you send an item to yourself, however, the sending player will be omitted.
+If the items are cheated in, no textbox will be displayed.
```
By cheated in I meant anything that has the server as the player that found it. So, yeah, start_inventory would count for that. I found it annoying to see all the pop-ups when using start_inventory, but perhaps this would just be confusing instead?

I don't feel strongly one way or the other, so if this is outside of the norm of what other worlds do then I'm happy to change it.

Exempt-Medic commented on worlds/saving_princess/Options.py:
```diff
@@ -0,0 +1,122 @@
+    display_name = "Screen Shake Intensity %"
+    range_start = 0
+    range_end = 100
+    default = 50
```
Oh, having played the game now, this no longer seems weird XD

LeonarthCG commented on worlds/saving_princess/docs/setup_en.md:
```diff
@@ -0,0 +1,120 @@
+If the game is unable to reconnect, save and restart.
+
+Although you can keep playing while disconnected, you won't get any items until you reconnect, not even items found in your own game.
+Once reconnected, however, all of your progress should sync up.
```
By "should sync up" I was trying to say that gm-apclientpp would take care of it.
I guess I should just say "will sync up"? Because if we assume the mod, the client and the protocol all work, then it will sync up, and if doesn't I don't think there's anything the user could really do to help it anyway?

The only way I've managed to make it not sync properly in my testing has been by wiping all of the local server's data and then starting it back up, but if it comes to that I'm not sure there's much a user can do about it.

Exempt-Medic replied:
Changing to "will" makes sense to me then

LeonarthCG commented on worlds/saving_princess/Client.py:
```diff
@@ -0,0 +1,215 @@
+            versions = json.load(versions_json)
+            versions[target_asset] = date
+    except (FileNotFoundError, KeyError, json.decoder.JSONDecodeError):
+        versions: json = {target_asset: date}
```
No, I don't think it is.
I honestly cannot remember why I did that.

LeonarthCG commented:
@Exempt-Medic 
> Something I forgot to comment on: The trap change is simply a flat chance, so if there are 10 filler items and 50% traps, there are not assuredly 5 traps, but there could be any number 0-10. Is this intended?

This is intentional, and I believe the docstring ("Likelihood of a filler item becoming a trap.") for the option is clear enough.

Is this design choice a big deal? I figured traps should be a surprise, and knowing the exact number of them kinda goes against that in my mind.

Exempt-Medic commented:
> @Exempt-Medic
> 
> > Something I forgot to comment on: The trap change is simply a flat chance, so if there are 10 filler items and 50% traps, there are not assuredly 5 traps, but there could be any number 0-10. Is this intended?
> 
> This is intentional, and I believe the docstring ("Likelihood of a filler item becoming a trap.") for the option is clear enough.
> 
> Is this design choice a big deal? I figured traps should be a surprise, and knowing the exact number of them kinda goes against that in my mind.

I don't think it's a big deal, just a design choice that different devs do wildly differently so i figured it was worth asking about just in case you hadn't considered it.

LeonarthCG commented on worlds/saving_princess/docs/setup_en.md:
```diff
@@ -0,0 +1,120 @@
+### Set up the connection details
+
+After launching the game, enter the Archipelago options menu through the in-game button with the Archipelago icon.
+From here, enter the different menus and type in the **server:port**, **slot**, as well as the **password**, if one is required.
```
This is what I came up with, I'm not sure if this is what you had in mind, so I'm asking before committing anything:

>After launching the game, enter the Archipelago options menu through the in-game button with the Archipelago icon.
>From here, enter the different menus and type in the following details in their respective fields:
>- **server:port** (e.g. `archipelago.gg:38281`)
>    * If hosting on the website, the server and port will be shown in your created room.
>- **slot name** (e.g. `Player`)
>    * This is your player name, which you chose along with your player options.
>- **password** (e.g. `123456`)
>   * If the room does not have a password, it should be left empty.
>
>This configuration persists through launches and even updates.

Exempt-Medic replied:
That looks great to me!

LeonarthCG commented on worlds/saving_princess/Rules.py:
```diff
@@ -0,0 +1,120 @@
+                        world.player
+                    ) and super_nice_check(state)
+
+    def is_power_on(state: CollectionState) -> bool:
```
If expanded pool is off, the expanded pool items do nothing. The game will just ignore them in favor of checking for the flags for the corresponding event (mimicking the vanilla game).

So, (if I have this right) using `state.has_any` here would make it so the generation doesn't think the player needs to reach `EVENT_ITEM_POWER_ON` to get power back if they have `EP_ITEM_POWER_ON` in their starting inventory.

Exempt-Medic replied:
Ah they get completely ignored, this makes sense then

LeonarthCG commented:
Unless I missed something, I believe I tackled all of the concerns.

I also removed the standalone launcher, leaving only the ArchipelagoLauncher one, and added option groups and presets.

Exempt-Medic reviewed (approved):
Recent changes LGTM, did 100 more singleplayer generations and 100 more 25-player generations. Tested playing through one game. All prior comments were addressed.

NewSoupVi reviewed (changes requested):
Incredibly nice implementation. I have one important request and two minor comments that aren't important.

NewSoupVi commented on worlds/saving_princess/__init__.py:
```diff
@@ -0,0 +1,174 @@
+from .Constants import *
+
+
+def launch_client():
+    from .Client import launch
+    launch_subprocess(launch, name=CLIENT_NAME)
+
+
+components.append(
+    Component(f"{GAME_NAME} Client", func=launch_client, component_type=ComponentType.CLIENT)
+)
```
Fellow Core Maintainer Black Sliver agrees with me that this is a good procedure for now.

LeonarthCG replied:
I think I should be able to implement passing the link to the game as well, so I'll look into that.
